helix-core/src/indent.rs

```diff
@@ -186,106 +188,338 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+    cursor: &mut TreeCursor<'a>,
+) -> (AddedIndent, AddedIndent) {
+    let (indent, next_indent) = scopes_contain_match(&query.indent, node, cursor);
+    let (outdent, next_outdent) = scopes_contain_match(&query.outdent, node, cursor);
+    let line = AddedIndent { indent, outdent };
+    let next = AddedIndent {
+        indent: next_indent,
+        outdent: next_outdent,
+    };
+    (line, next)
+}
 
-        if query.outdent.contains(node_kind) {
-            increment_from_line -= 1;
+/// Use the syntax tree to determine the indentation for a given position.
+/// This can be used in 2 ways:
+///
+/// - To get the correct indentation for an existing line (new_line=false), not necessarily equal to the current indentation.
+///   - In this case, pos should be inside the first tree-sitter node on that line.
+///     In most cases, this can just be the first non-whitespace on that line.
+///   - To get the indentation for a new line (new_line=true). This behaves like the first usecase if the part of the current line
+///     after pos were moved to a new line.
+///
+/// The indentation is determined by traversing all the tree-sitter nodes containing the position.
+/// Each of these nodes produces some [AddedIndent] for:
+///
+/// - The line of the (beginning of the) node. This is defined by the scope `all` if this is the first node on its line.
+/// - The line after the node. This is defined by:
+///   - The scope `tail`.
+///   - The scope `all` if this node is not the first node on its line.
+/// Intuitively, `all` applies to everything contained in this node while `tail` applies to everything except for the first line of the node.
+/// The indents from different nodes for the same line are then combined.
+/// The [IndentResult] is simply the sum of the [AddedIndent] for all lines.
+///
+/// Specifying which line exactly an [AddedIndent] applies to is important because indents on the same line combine differently than indents on different lines:
+/// ```ignore
+/// some_function(|| {
+///     // Both the function parameters as well as the contained block should be indented.
+///     // Because they are on the same line, this only yields one indent level
+/// });
+/// ```
+///
+/// ```ignore
+/// some_function(
+///     parm1,
+///     || {
+///         // Here we get 2 indent levels because the 'parameters' and the 'block' node begin on different lines
+///     },
+/// );
+/// ```
+fn treesitter_indent_for_pos(
+    query: &IndentQuery,
+    syntax: &Syntax,
+    indent_style: &IndentStyle,
+    text: RopeSlice,
+    line: usize,
+    pos: usize,
+    new_line: bool,
+) -> Option<String> {
+    let mut cursor = syntax.tree().walk();
+    let byte_pos = text.char_to_byte(pos);
+    let mut node = match get_lowest_node(syntax.tree().root_node(), query, byte_pos) {
+        Some(n) => n,
+        None => {
+            return None;
         }
-        if query.indent.contains(node_kind) {
-            increment_from_line += 1;
+    };
```

Comment:
This can just use `?`

Reply:
Nice, I didn't know the `?` operator could be used with `Option` too.